tsdb/shard.go

```diff
@@ -723,6 +724,39 @@ func (s *Shard) validateSeriesAndFields(points []models.Point, tracker StatsTrac
 	return points[:j], fieldsToCreate, err
 }
 
+const unPrintReplRune = '?'
+const unPrintMaxReplRune = 3
+
+// makePrintable - replace invalid and non-printable unicode characters with a few '?' runes
+func makePrintable(s string) string {
+	b := strings.Builder{}
+	b.Grow(len(s))
+	for _, r := range strings.ToValidUTF8(s, string(unPrintReplRune)) {
+		if !unicode.IsPrint(r) || r == unicode.ReplacementChar {
+			b.WriteRune(unPrintReplRune)
```

Comment:
If you kept the 3-byte counter here instead, you wouldn't compress valid unicode `WHAT????` to `WHAT???`

Reply:
I also want to elide runs of '?' generated by `string.ToValidUTF8`, which in the customer scenario are in the hundreds of characters range (they put thousand-character runs of invalid Unicode characters in their line protocol).  The only complete solution for perfect operation I saw was to re-implement `strings.ToValidUTF8` with run-length detection, and that seemed overkill.

Reply:
Now fixed with `unicode.ReplacementChar` and a single replacement elision loop.